`PartialDeep`: Ignore `Date` and `RegExp` types

sdotson opened the pull request:
Addresses issue https://github.com/sindresorhus/type-fest/issues/207, using feedback/code from closed pr https://github.com/sindresorhus/type-fest/pull/208 as a starting point.

This pr adds an additional `ExceptTypes` argument to `PartialDeep`, allowing consumers to state that certain types like `Date` or perhaps custom types written by somebody else, should be treated as exceptions.

Super excited to contribute and open to whatever feedback. Glad to help.

sindresorhus commented:
I think it should always ignore `Date | RegExp`. So instead of extending `Primitive`. Make a new local type called `Builtins` and make it `Primitive | Date | RegExp`.

sindresorhus commented on source/partial-deep.d.ts:
```diff
@@ -33,49 +33,55 @@ settings = applySavedSettings({textEditor: {fontWeight: 500}});
+	[KeyType in keyof ObjectType]?: PartialDeep<ObjectType[KeyType], ExceptTypes>
 };
+
+type AddSymbolToPrimitive<T> = T extends
```
This is unused

sdotson commented:
@sindresorhus Thank you for the feedback. I've created the `BuiltIns` as you suggested, removed the `ExceptionTypes` parameter, and updated the tests.

sindresorhus commented on source/partial-deep.d.ts:
```diff
@@ -33,8 +33,8 @@ settings = applySavedSettings({textEditor: {fontWeight: 500}});
 @category Map
 */
-export type PartialDeep<T> = T extends Primitive
-	? Partial<T>
```
Intentional that you're removing the `Partial`?

sdotson replied:
Yes. Couldn't really get `RegExp` and `Date` to work with it there. Are there any drawbacks to removing it? I'll admit I don't fully understand the advantages of doing `Partial` with primitives and `RegExp`/`Date`.

sindresorhus replied:
> Are there any drawbacks to removing it?

I don't know. I recommend using Git blame to figure out who added it and then mentioning them here.

sdotson replied:
@kainiedziela I see you made the most recent change to this, which was actually adding this 2 years ago. Any concerns?

kainiedziela replied:
No concerns

sindresorhus commented:
In case you're interested, we should apply this fix to the other deep types too: https://github.com/sindresorhus/type-fest/issues/347
